Latest WebAssembly ESM Integration rebase

guybedford opened the pull request:
<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial. 
When editing this comment after the PR is created, check that PR-Preview doesn't overwrite your changes.
If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

WebAssembly ESM Integration including Source Phase Imports

This represents the latest rebase of the ESM Integration [spec](https://github.com/WebAssembly/esm-integration), replacing the original PR in https://github.com/whatwg/html/pull/4372. Major changes since then include [source phase imports](https://github.com/tc39/proposal-source-phase-imports) support and instantiation being updated to be synchronous.

There is an early validation error algorithm that is applied when modules are created, which needed to be updated to support the source phase imports proposal. Specifically, constructed modules may not have their dependencies loaded when in the source phase. This validation logic is now located in HostLoadImportedModule, and performed against all referrer dependencies on the first call to HostLoadImportedModule for a given referrer, detected by the first argument matching the first module of the Cyclic Module Record (indicating the module is loading its dependencies).

Since the ESM Integration specification is still in Phase 3, its specification is referenced directly here. As soon as the proposal is upstreamed we can reference the main WebAssembly JS API specification.

- [x] At least two implementers are interested (and none opposed):
   * Chromium (with V8 source phase [design doc](https://docs.google.com/document/d/1yetUlxl_yHk0ooT-hg9ylJeQ6Q77_Ttm777tHbclwqg/edit))
   * Node.js
   * Deno
   * Previously WebKit, pending impl refresh
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * Initial ESM Integration: https://github.com/web-platform-tests/wpt/pull/32862
   * Source phase support: https://github.com/web-platform-tests/wpt/pull/42467, https://github.com/web-platform-tests/wpt/pull/45401
   * Tentative -> Non-tentative PR: https://github.com/web-platform-tests/wpt/pull/46556
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://issues.chromium.org/issues/40244733, with source phase support tracked in https://issues.chromium.org/issues/42204365
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1899832
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=274908
   * Deno (only for timers, structured clone, base64 utils, channel messaging, module resolution, web workers, and web storage): https://github.com/denoland/deno/issues/2552
   * Node.js (only for timers, structured clone, base64 utils, channel messaging, and module resolution): https://github.com/nodejs/node/issues/53178
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/mdn/issues/549
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10380/acknowledgements.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">/acknowledgements.html</a>  ( <a href="https://whatpr.org/html/10380/9a01ce8...3868460/acknowledgements.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">diff</a> )
<a href="https://whatpr.org/html/10380/indices.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">/indices.html</a>  ( <a href="https://whatpr.org/html/10380/9a01ce8...3868460/indices.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">diff</a> )
<a href="https://whatpr.org/html/10380/infrastructure.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10380/9a01ce8...3868460/infrastructure.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">diff</a> )
<a href="https://whatpr.org/html/10380/references.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">/references.html</a>  ( <a href="https://whatpr.org/html/10380/9a01ce8...3868460/references.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">diff</a> )
<a href="https://whatpr.org/html/10380/webappapis.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">/webappapis.html</a>  ( <a href="https://whatpr.org/html/10380/9a01ce8...3868460/webappapis.html" title="Last updated on Sep 4, 2024, 6:40 AM UTC (3868460)">diff</a> )

annevk commented on source:
```diff
@@ -106980,21 +106998,38 @@ document.querySelector("button").addEventListener("click", bound);
-     <var>moduleType</var> is "<code data-x="">json</code>", then set <var>moduleScript</var> to
-     the result of <span>creating a JSON module script</span> given <var>sourceText</var> and
-     <var>settingsObject</var>.</p></li>
+     <li><p>If <var>mimeType</var>'s <span data-x="MIME type essence">essence</span> is
```
If an `<li>` contains multiple children the `<p>` should be on the next line and indented.

annevk commented on source:
```diff
@@ -106980,21 +106998,38 @@ document.querySelector("button").addEventListener("click", bound);
-     the result of <span>creating a JSON module script</span> given <var>sourceText</var> and
-     <var>settingsObject</var>.</p></li>
+     <li><p>If <var>mimeType</var>'s <span data-x="MIME type essence">essence</span> is
+     <code>application/wasm</code> and <var>moduleType</var> is
```
Missing quotation marks around the string.

annevk commented on source:
```diff
@@ -109090,7 +109209,7 @@ dictionary <dfn dictionary>PromiseRejectionEventInit</dfn> : <span>EventInit</sp
   The <span>URL record</span> is the <span data-x="concept-request-url">request URL</span> at which
   the module was fetched, and the <span>string</span> indicates the type of the module (e.g. "<code
-  data-x="">javascript</code>").  The <span>module map</span>'s values are either a <span>module
+  data-x="">javascript-or-wasm</code>").  The <span>module map</span>'s values are either a <span>module
```
```suggestion
  data-x="">javascript-or-wasm</code>"). The <span>module map</span>'s values are either a <span>module
```

annevk replied:
But also, this requires rewrapping as it now exceeds a 100 columns.

annevk commented on source:
```diff
@@ -109130,7 +109249,7 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
    <p>Since module type is also part of the <span>module map</span> key, the following code will
    create two separate entries in the <span>module map</span> (the type is "<code
-   data-x="">javascript</code>" for the first, and "<code data-x="">css</code>" for the second):
+   data-x="">javascript-or-wasm</code>" for the first, and "<code data-x="">css</code>" for the second):
```
This also seems too long.

guybedford commented:
Thanks for the review, I've addressed the changes in https://github.com/whatwg/html/pull/10380/commits/f816de8a0288b58a01b26796aea1b878d1b51087.

annevk commented on source:
```diff
@@ -107137,21 +107155,39 @@ document.querySelector("button").addEventListener("click", bound);
+       <span data-x="concept-response-url">URL</span>, and <var>options</var>.</p></li>
+      </ol>
+     </li>
+     <li>
```
Newline before `<li>` for consistency.

annevk commented on source:
```diff
@@ -107381,6 +107417,90 @@ document.querySelector("button").addEventListener("click", bound);
+    <ol>
+     <li><p>Set <var>script</var>'s <span data-x="concept-script-parse-error">parse error</span> to
+     <var>error</var>.</p></li>
+     <li><p>Return <var>script</var>.</p></li>
```
Newline before `<li>` for consistency.

annevk reviewed (approved):
@nicolo-ribaudo could you have a look at this please? Looks editorially okay to me.

nicolo-ribaudo commented on source:
```diff
@@ -107381,6 +107417,90 @@ document.querySelector("button").addEventListener("click", bound);
+  <ol>
+   <li><p>If <span data-x="concept-environment-noscript">scripting is disabled</span> for
+   <var>settings</var>, then set <var>bodyBytes</var> to the byte sequence 0x00 0x61 0x73 0x6d 0x01
+   0x00 0x00 0x00, corresponding to an empty WebAssembly module with only the magic bytes and version
+   number provided.</p></li>
```
"corresponding to an empty WebAssembly module with only the magic bytes and number provided" should probably go in a `<p class="node">`.

nicolo-ribaudo commented on source:
```diff
@@ -107381,6 +107417,90 @@ document.querySelector("button").addEventListener("click", bound);
+      </ol>
+     </li>
+
+     <li><p>Let <var>moduleType</var> be the result of running the <span>module type from module
+     request</span> steps given <var>requested</var>.</p></li>
+     <li><p><span>Assert</span>: <var>moduleType</var> is "<code data-x="">javascript-or-wasm</code>"
+     </p></li>
```
Given that we never actually use this `moduleType` other than in the note, we can delete these steps and reword the note to
```js
WebAssembly modules do not define attributes for their imports, therefore the <span data-x="module type from request">module type from<span> any of their [[ModuleRequest]]s will always be "<code data-x="">javascript-or-wasm</code>".
```

nicolo-ribaudo commented on source:
```diff
@@ -107455,21 +107575,21 @@ document.querySelector("button").addEventListener("click", bound);
       <var>moduleType</var> to null.</p>
 
-      <p class="note">This specification uses the "<code data-x="">javascript</code>" module type
+      <p class="note">This specification uses the "<code data-x="">javascript-or-wasm</code>" module type
       internally for <span data-x="JavaScript module script">JavaScript module scripts</span>, so
```
_or WebAssembly module scripts_

nicolo-ribaudo commented on source:
```diff
@@ -107381,6 +107418,91 @@ document.querySelector("button").addEventListener("click", bound);
+     <li>
+      <p class="note">WebAssembly modules do not define attributes for their imports, therefore the
+      <var>moduleType</var> will always be "<code data-x="">javascript-or-wasm</code>".</p>
+     </li>
```
This accidentally came back after rebasing

guybedford replied:
Still working through the commits, give me 2 mins.

guybedford commented:
I've integrated the latest changes here so we should be good to go.

Just let me know if I can do anything to help get the participation check passing.

domenic commented:
> Just let me know if I can do anything to help get the participation check passing.

It looks like you signed up as an individual at some point, but you work for Fastly, which has signed up as an entity. However, Fastly has only signed up to contribute to Fetch.

So this is a bit complicated. The ideal steps needed are:

- @fastlyneal needs to send a PR to update [this part of entities.json](https://github.com/whatwg/participant-data/blob/4927d0082501b85259e0af0db120e3c4ae5ee63c/entities.json#L518-L539) to include the HTML workstream. (He might also want to remove mnot as a GitHub contact there, since mnot appears to no longer work at Fastly...)
- You can send a PR to remove yoursself from https://github.com/whatwg/participant-data/blob/main/individuals.json
- You need to join, or make public your membership in, the `Fastly` GitHub organization.

guybedford commented:
The participation check is now passing. @domenic thanks for the clear help with this.

domenic commented:
This looks ready to merge to me, but I'm not 100% sure that we have multiple browsers interested in implementing exactly this shape of the proposal. We have zero browsers passing the tests, for example: https://wpt.fyi/results/wasm/webapi/esm-integration?label=experimental&label=master&aligned

annevk commented:
Although the WebKit team at Apple was initially somewhat skeptical about bundling JS and Wasm as a single destination, we're okay with that now under the expectation that the privileges and capabilities of JS and Wasm won't diverge.

Who can speak for Chromium, maybe @syg?

And Gecko would be @codehag or @mgaudet perhaps?

codehag commented:
This looks ok from our side.

guybedford commented:
@ajklein may also be able to speak to the current progress on the V8 implementation here.

dminor commented:
> Although the WebKit team at Apple was initially somewhat skeptical about bundling JS and Wasm as a single destination, we're okay with that now under the expectation that the privileges and capabilities of JS and Wasm won't diverge.
> 
> Who can speak for Chromium, maybe @syg?
> 
> And Gecko would be @codehag or @mgaudet perhaps?

Anne, feel free to ping me on things like this as well :)

ajklein commented:
Speaking to implementation progress: in V8/Chromium, we are currently working on supporting source phase imports for loading WebAssembly modules as source (tracked in https://issues.chromium.org/issues/42204365). We are not currently prioritizing work on full ESM/Wasm integration. Essentially, we're working on item (1) of the Wasm ESM Integration proposal's [Progressive Implementation Support](https://github.com/WebAssembly/esm-integration/blob/main/proposals/esm-integration/README.md#progressive-implementation-support) section.

syg commented:
IIUC, this PR adds Wasm support for both source phase imports (e.g. `import source foo from 'bar.wasm'`) and evaluation phase imports (e.g. `import 'bar.wasm'`). As @ajklein says in https://github.com/whatwg/html/pull/10380#issuecomment-2269980165, the V8/Chromium has planned work on source phase imports _only_.

It's also my understanding that the WPT tests linked above test both source phase imports and evaluation phase imports.

@annevk Is WebKit planning to implement and ship Wasm evaluation phase imports as well as source phase imports?

@codehag What about Gecko?

If the answer to both questions above is yes, then seems fine. Otherwise I feel that we should separate the source phase support from the evaluation phase support, as we shouldn't merge something that won't (fully) reflect reality for a while.

codehag commented:
Thanks for the ping @syg. My understanding is there are two parts to your question:

1) Will Gecko Support both source phase imports and evaluation phase imports for wasm?

If they are both standardized, our plan is to support both. The original proposal was to only have evaluation phase imports, and source phase imports are a new thing. That brings us to ...

2) given the changes to the spec in recent years, source phase imports make more sense for wasm. Should we support both  source phase imports and evaluation phase imports _right now_, or should we wait on evaluation imports to see if they have a use case?

Most people importing wasm today are doing so as "source" first, and then instantiating it themselves. This was a major motivation for source phase imports to begin with. This begs the question: If evaluation phase imports are specified, will they actually be used? Or are we supporting them purely out of spec inertia?

Consistency between JS and WASM importing would be nice, especially since they will be using the same syntax. Evaluation phase imports have been [supported by webkit behind a flag ](https://bugs.webkit.org/show_bug.cgi?id=236268) for some time IIUC. I could see some users having potential confusion if imports between WASM and JS are not equivalent. Maybe @guybedford could speak to whether evaluation phase is important for users? 

I don't have a concrete answer for this question. We could co either way.

syg commented:
Thanks, @codehag. The "we will support if standardized" is a bit entangled so let me try to disentangle that AFAIU:

- Evaluation phase Wasm ESM integration is not yet standardized / ready to implement. Whether evaluation phase Wasm ESM integration is standardized is being done in Wasm CG, not WhatWG.
- If Wasm CG standardizes evaluation phase imports, we need HTML support.
- Source phase Wasm ESM integration is already Stage 3 in TC39.
- We need HTML support for source phase imports.

IOW, my understanding of this HTML PR is purely to support proposals standardized elsewhere. We don't really need to get into the actual motivations for the other proposals here, there're other venues for that already. But because this PR is purely about support, I want it to reflect the most likely interoperable future.

Best I can tell, source phase imports is definitely in that future. I don't know if evaluation phase imports is in that future yet. Chrome isn't working on it right now, thus the question to Mozilla and Apple if you have a different opinion.

guybedford commented:
While the motivation for source phase imports remains that it is considered the more useful route for the instantiation of WebAssembly modules, this does not negate the need for or usefulness of the instance / evaluation phase even if those are considered a much smaller use case at this point. There are many packages on npm today that completely work with standards-compatible ESM evaluation phase integration.

There are no plans to deprecate or remove the evaluation phase. The standardization of the ESM Integration spec at the Wasm CG into Phase 3 included both phases, with the original spec mechanics and WPT tests for the evaluation phase remaining throughout the source phase process per the Webkit implementation, with just a couple of small changes (mainly moving to synchronous instantiation). Per the README note linked, the weakening of the implementation constraint to allow shipping one without the other specifically does not indicate that either should not ever be implemented - the standard includes both phases and that is what is expected to be implemented.

Node.js ships the instance phase under the `--experimental-wasm-modules` flag already. For Node.js, our plan once V8 source phase syntax is shipped in Node.js is to then unflag with both the source and evaluation phase supported.

As @syg points out, the conformance decision here is entirely made in the ESM Integration spec though. We could possibly add some further clarifications on the HTML side to note both phase are specified if that would be useful.

syg commented:
> Yes, this PR adds HTML support for wasm imports regardless of their phase (source vs evaluation). Which phases are supported is entirely controlled by the Wasm proposal, which defines the WebAsssembly Module Record.

This was part of my question: is it exactly the same? That is, if you were to limit this proposal to just supporting source phase imports, can you remove stuff from this PR? E.g. this PR talks about module requests for WebAssembly Module Scripts, which isn't needed for source phase imports, right?

nicolo-ribaudo commented:
That step there also affects `import source`. i.e. this will result in an error (assuming that there is no import map mapping `"not a valid URL"` to something):
```js
import source foo from "./x.wasm"
```
```wasm
(module
  (import "not a valid URL" "num" i32)
)
```

That however raises indeed a question: is this intentional? @guybedford

nicolo-ribaudo commented:
I talked a bit with Guy about that, and no that is definitely not intentional. Doing `import source x from "foo"` should not validate `"foo"`'s imports (regardless of whether `foo` is a WASM source, of a JS source if/when we'll add them).

The reason HTML has that check is that we want to abort the loading process as early as possible when we know that it will fail: in case of
```js
import "./valid.js";
import "invalid specifier";
```
we want to avoid starting the network request for `./valid.js`, since we can easily tell that the process will fail anyway.

It seems like we can get the same behavior by moving those from "create a XXX module script" to the beginning of `HostLoadImportedModule`, validating `referrer.[[RequestedModules]]` before doing anything else. There is no observable work happening between when "create a JavaScript module script" returns and when the first `HostLoadImportedModule` call for its dependencies happen, so the only difference is that we'd do this pre-validation only when we know that we want to load the dependencies of that module.

guybedford commented:
Thanks @nicolo-ribaudo for bringing up this extra validation step - as we discussed, I've added a refactoring in https://github.com/whatwg/html/pull/10380/commits/da7425ed678648d18dc17ad517e5410ac98868d4 to move the validation into HostLoadImportedModule so that source phase imports no longer validate specifier resolution since they are supposed to remain virtual.

> That is, if you were to limit this proposal to just supporting source phase imports, can you remove stuff from this PR? E.g. this PR talks about module requests for WebAssembly Module Scripts, which isn't needed for source phase imports, right?

@syg this led to a great catch - with the latest validation refactoring which can be considered a simplification in only performing validation and resolve validation once during the main load algorithm, yes this PR now remains mostly equivalent with and without the source phase integrated on the Wasm spec side. I've also added an additional note in the last commit to mention the source phase explicitly and to avoid any ambiguity in https://github.com/whatwg/html/pull/10380/commits/6a5cc63eca2a27305edbdfeecd3df492227057c8, let me know if that helps?

nicolo-ribaudo commented on source:
```diff
@@ -109540,6 +109573,44 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
 
   <ol>
+   <li id="validate-requested-module-specifier">
+    <p>If <var>moduleRequest</var>.[[Attributes]] contains a <span>Record</span> <var>entry</var>
```
To preserve the existing behavior, we need to apply this validation to
```html
    <p><span data-x="list iterate">For each</span> <span>ModuleRequest record</span>
    <var>requested</var> of <var>referrer</var>.[[RequestedModules]]:</p>
```

so that in case of
```js
import "./x.js"
import "<invalid>"
```
we skip work related to `x.js`

guybedford replied:
This validation is already done before calling this function though in InnerModuleLoading at https://ci.tc39.es/preview/tc39/ecma262/sha/282c013e43978ae23561eeb9caa13b2b9547ff7d/#sec-InnerModuleLoading, so I'm wondering if we really still need to be duplicating this logic in HTML given it's clearly defined on the ECMA-262 side?

annevk commented:
If you edit OP it makes PR Preview run again. I just did that and it succeeded.

syg commented:
I'm not too familiar with the HTML side of the module loading machinery but I think this makes mostly sense to me now, and I'm happy with it landing. Recapping my understanding of the state of the world:

1. Wasm ESM integration is specced across three (3) specs: HTML, ecma262, and the [ESM integration proposal](https://webassembly.github.io/esm-integration/js-api/index.html#esm-integration).
1. This PR adds the ability to fetch and parse Wasm modules. With the HostLoadImportedModule refactoring, it is agnostic to the phase of the import.
1. What phase a module is imported with is entirely driven from within ecma262 and JavaScript programs.
1. The ESM integration proposal specs the behavior for both source and evaluation phases.
1. Chrome isn't working on evaluation time Wasm module support at this time.
1. I haven't heard any commitment from Safari or Firefox for evaluation time Wasm module support.

Given the points (5) and (6) above, I'd prefer the informative note in [this section](https://webassembly.github.io/esm-integration/js-api/index.html#esm-integration) of the wasm spec to be more reflective of near- (maybe medium-?) term web reality, but that's nothing to do with this PR.

syg commented on source:
```diff
@@ -107344,70 +107344,6 @@ document.querySelector("button").addEventListener("click", bound);
-    <p class="note">This step is essentially validating all of the requested module specifiers and
-    type attributes. We treat a module with unresolvable module specifiers or unsupported type
-    attributes the same as one that cannot be parsed; in both cases, a syntactic issue makes it
-    impossible to ever contemplate linking the module later.</p>
```
Is this note not worth preserving?

guybedford replied:
After discussing with @nicolo-ribaudo again, I was able to figure out the correct refactoring, and have also clarified the note a little more.

guybedford commented:
@syg thanks for your summary here, that sounds like a good outline to me.

With regards to:

> Given the points (5) and (6) above, I'd prefer the informative note in [this section](https://webassembly.github.io/esm-integration/js-api/index.html#esm-integration) of the wasm spec to be more reflective of near- (maybe medium-?) term web reality, but that's nothing to do with this PR.

I've added a possible adjustment to this note in https://github.com/WebAssembly/esm-integration/pull/91, and feel free to comment there further.

Just waiting for a final review from @nicolo-ribaudo then we should have all reviews completed for this to land.

nicolo-ribaudo commented on source:
```diff
@@ -109540,6 +109573,23 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+   request</span> steps given <var>moduleRequest</var>.</p></li>
+
+   <li>
+    <p>If the result of running the <span>module type allowed</span> steps given
```
This should be moved back inside the loop as it was before this PR. Otherwise, this example changes from throwing a ReferenceError to throwing a SyntaxError:
```js
import "a";
import "b" with { type: "invalid" }
import "c" with { invalid: "foo" }
```

guybedford replied:
Thanks, updated.

nicolo-ribaudo commented on source:
```diff
@@ -109601,6 +109651,40 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+        such that <var>entry</var>.[[Key]] is not "<code data-x="">type</code>", then:</p>
+
+        <ol>
+         <li><p>Let <var>error</var> be a new <code>SyntaxError</code> exception.</p></li>
+
+         <li><p>Set <var>script</var>'s <span data-x="concept-script-parse-error">parse error</span>
+         to <var>error</var>.</p></li>
+
+         <li><p>Return <var>script</var>.</p></li>
```
Since now we are using an already-created script, and since HostLoadImportedModule does not have a return value, this should not be
```
1. Let completion be Completion Record { [[Type]]: throw, [[Value]]: a new SyntaxError, [[Target]]: empty }.
2. Perform FinishLoadingImportedModule( referrer , moduleRequest , payload , completion ).
3. Return.
```

guybedford replied:
Done.

nicolo-ribaudo reviewed (commented):
This now looks good to me, I'm somewhat confident that it preserve all the existing behavior.

Btw, there are some problems with the Wattsi server. Yesterday I was going to send a long review explaining how the changes were still wrong because the Wattsi preview was stuck at an old commit rather than at the last one that passed the build.

domenic reviewed (commented):
I pushed a few nits; please check that they didn't change anything.

I have two questions:

- I'm confused on the name of the [WASMESM] specification. It calls itself "WebAssembly JavaScript Interface" in the heading of https://webassembly.github.io/esm-integration/js-api/index.html . But that name is already taken by our [WASMJS] reference, https://webassembly.github.io/spec/js-api/ ? This PR seems to call it "WebAssembly JavaScript Module Integration" but I cannot find any evidence of that being an official referenceable name.

- Can you summarize why we relocated all the dependency checking to HostResolveImportedModule? That's a large change.

Additionally, if you could update the OP with as helpful a commit message as possible, that would be great. Explaining what exactly changes, what the boundary is between the two specs, maybe the different import phases.

guybedford commented:
@domenic nits all look good to me. I do have a slight preference for `Module Record` over `Cyclic Module Record` terminology, but being clear works just as good. I've updated the top-line of the PR comment to include the answers to your questions, let me know if that answers things for you?

If there's a better way to handle the Phase 3 status of the ESM Integration in references, I'm open to alternatives.

domenic commented:
> If there's a better way to handle the Phase 3 status of the ESM Integration in references, I'm open to alternatives.

Could you give the document a different title, instead of having two documents with the same title?

guybedford commented:
Sure, I've pushed a commit to update the ESM integration reference to be the spec repo instead of the spec document, and set the title as the direct proposal's name.

domenic commented:
Sorry, that wasn't what I meant; I meant, can you update the spec document? I'm just very surprised the WASM CG is putting out two specs with the same name but different contents, and am unsure whether we should be referencing WASM CG documents if that's their status.

guybedford commented:
Okay, reverted the last commit and it seems the exceptions proposal did similar here previously in altering the title to include the proposal name. I've landed the title update in https://github.com/WebAssembly/esm-integration/pull/93, but still need to get the spec document to update properly through CI, so will fix that tomorrow and ping back here then. Let me know if there's anything else, and thanks for the reviews!

guybedford commented:
@domenic the title on the spec is now updated, so we should finally be there now.

nicolo-ribaudo commented:
I don't see a comment answering

> Can you summarize why we relocated all the dependency checking to HostResolveImportedModule? That's a large change.

The summary is that we can split module loading in two phases:
1. Load a module
2. Load its dependencies

The first one ends at the end of the "create a X module" steps, and the second one is in HostLoadImportedModule.

Somewhere between 1 and 2 we need to validate those dependencies. Whether it was at the end of 1 or at the beginning of 2 was originally not observable. With source phase imports it is observable, because we can stop after 1, and we only want the validation to happen when we know that 2 is going to happen.

LePichu commented:
I know I am kinda late but I haven't seen this anywhere, shouldn't this work to interop work on Import Assertions too in someway? Historically bundlers have allowed you to bundle binary blobs like PNGs, and WASM blobs inline, and text data like JSON too, Import Assertions were supposed to allow for better readability and trying to make a standard way of importing non-JS modules into JS, shouldn't this do something like `import { foo } from "./bar.wasm" with { type: "wasm" }` instead?

nicolo-ribaudo commented:
This has been discussed multiple times — a goal here is making it possible to swap JS modules with wasm modules in a transparent way.

The goal of import attributes on the web was security: JSON and CSS have less capabilities than JS, and the attribute is there to make sure that somebody doesn't try to attack you by providing a JS file where you expect a JSON one. Wasm instead can import JS, and it runs at the same privilege level of JS.

This is similar to how there is no `type: "JS"`.

Note that bundlers can decide to support their own custom attributes, as long as they strip them away when bundling.
